File: apiome-ui/src/app/components/ade/dashboard/repositories/repositoryConflictPolicy.ts
/**
 * Repository refresh conflict policy — types and presentation rules (RAR-4.5, #3531).
 *
 * When auto-refresh re-imports a changed file it can find that the version the original
 * import produced has been hand-edited in Apiome since. The divergence guard (RAR-4.4)
 * detects that; this policy decides what happens next. Three answers, and the difference
 * between them is the difference between "we lost an afternoon of edits" and "we shipped a
 * stale spec", so the screen has to say what each one *does*, not just name it.
 *
 * Two rules live here rather than in the component:
 *
 *  * {@link effectivePolicyFor} — the same `file -> repository -> default` precedence the
 *    server applies, so the panel can show a file's real policy without asking again.
 *  * {@link POLICY_COPY} — one place for the label, the one-line consequence, and the tone,
 *    so the selector, the override rows and the tests all read the same words.
 *
 * Deliberately React-free, so every rule is unit-testable without a DOM — the same split as
 * `repositoryWebhookIpAllowlist.ts` and `repositoryQuotaTelemetry.ts` next to it.
 */

/** The three policies the API accepts; the tokens match the stored values exactly. */
export type ConflictPolicy = 'overwrite' | 'hold-for-review' | 'new-branch';

/** The policy in force when neither the file nor the repository sets one. */
export const DEFAULT_CONFLICT_POLICY: ConflictPolicy = 'hold-for-review';

/** Every policy, in the order the selector renders them (safest default in the middle). */
export const CONFLICT_POLICIES: readonly ConflictPolicy[] = [
  'hold-for-review',
  'overwrite',
  'new-branch',
] as const;

/** One per-file exception to the repository-wide policy. */
export interface ConflictPolicyOverride {
  branch: string;
  path: string;
  policy: ConflictPolicy;
  createdBy: string | null;
  createdAt: string | null;
  updatedAt: string | null;
}

/** What `GET /api/repositories/{id}/conflict-policy` returns. */
export interface ConflictPolicyResponse {
  success: boolean;
  repositoryId: string;
  policy: ConflictPolicy;
  defaultPolicy: ConflictPolicy;
  availablePolicies: ConflictPolicy[];
  overrides: ConflictPolicyOverride[];
  error?: string;
}

/** Where an applied policy came from, for the "inherits from repository" copy. */
export type ConflictPolicySource = 'file' | 'repository' | 'default';

/**
 * Label, consequence and tone for each policy.
 *
 * `tone` is what the row is drawn with: `overwrite` is the only policy that can lose work
 * that is not in the repository, so it is the only one drawn as a warning.
 */
export const POLICY_COPY: Record<
  ConflictPolicy,
  { label: string; detail: string; tone: 'good' | 'neutral' | 'warn' }
> = {
  'hold-for-review': {
    label: 'Hold for review',
    detail:
      'The refresh is skipped and the file is flagged as diverged. Nothing is overwritten until someone resolves it.',
    tone: 'good',
  },
  overwrite: {
    label: 'Overwrite',
    detail:
      'The refresh wins: the repository version replaces the edits made in Apiome. The divergence is still recorded.',
    tone: 'warn',
  },
  'new-branch': {
    label: 'New branch',
    detail:
      'The current version is left untouched and the refresh lands on a new branch, so neither the edit nor the upstream change is lost.',
    tone: 'neutral',
  },
};

/** Panel/badge accent per tone, so the consequence reads before the text does. */
export const POLICY_TONE_CLASSES: Record<'good' | 'neutral' | 'warn', string> = {
  good: 'border-emerald-200 text-emerald-700 dark:border-emerald-800 dark:text-emerald-300',
  neutral: 'border-gray-200 text-gray-700 dark:border-gray-700 dark:text-gray-300',
  warn: 'border-amber-300 text-amber-800 dark:border-amber-700 dark:text-amber-300',
};

/**
 * Narrow an arbitrary value to a {@link ConflictPolicy}.
 *
 * Anything unrecognised — including null, undefined and a token from a newer server —
 * yields `fallback`, which mirrors the server's rule that an unusable value means "not set"
 * rather than an error. Because the broadest fallback is hold-for-review, every degradation
 * path is the safe one.
 *
 * @param value Raw value from an API payload or a form control.
 * @param fallback What to return when `value` is not a known policy.
 * @returns The narrowed policy.
 */
export function asConflictPolicy(
  value: unknown,
  fallback: ConflictPolicy = DEFAULT_CONFLICT_POLICY
): ConflictPolicy {
  return typeof value === 'string' && (CONFLICT_POLICIES as readonly string[]).includes(value)
    ? (value as ConflictPolicy)
    : fallback;
}

/**
 * Resolve the policy in force for one file, and say where it came from.
 *
 * Applies the server's precedence — `per-file override -> repository policy -> default` —
 * so the panel can label a row "inherits from repository" without a second request.
 *
 * @param path Repository-relative path of the file.
 * @param branch The branch the file was imported from.
 * @param repositoryPolicy The repository-wide policy.
 * @param overrides Every per-file override the repository has.
 * @returns The applied policy and its source.
 */
export function effectivePolicyFor(
  path: string,
  branch: string,
  repositoryPolicy: ConflictPolicy,
  overrides: readonly ConflictPolicyOverride[]
): { policy: ConflictPolicy; source: ConflictPolicySource } {
  const match = overrides.find((o) => o.path === path && o.branch === branch);
  if (match) return { policy: asConflictPolicy(match.policy), source: 'file' };
  return { policy: repositoryPolicy, source: 'repository' };
}

/**
 * Parse a conflict-policy API payload into the panel's shape.
 *
 * Every field is narrowed rather than trusted: a malformed payload renders the safe default
 * with no overrides instead of throwing inside a render.
 *
 * @param raw The parsed JSON body.
 * @returns The normalised response.
 */
export function parseConflictPolicyResponse(raw: unknown): ConflictPolicyResponse {
  const body = (raw ?? {}) as Record<string, unknown>;
  const policy = asConflictPolicy(body.policy);
  const overridesRaw = Array.isArray(body.overrides) ? body.overrides : [];
  return {
    success: body.success !== false,
    repositoryId: typeof body.repositoryId === 'string' ? body.repositoryId : '',
    policy,
    defaultPolicy: asConflictPolicy(body.defaultPolicy),
    availablePolicies: Array.isArray(body.availablePolicies)
      ? body.availablePolicies.map((p) => asConflictPolicy(p))
      : [...CONFLICT_POLICIES],
    overrides: overridesRaw.map((o) => {
      const row = (o ?? {}) as Record<string, unknown>;
      return {
        branch: typeof row.branch === 'string' ? row.branch : '',
        path: typeof row.path === 'string' ? row.path : '',
        policy: asConflictPolicy(row.policy),
        createdBy: typeof row.createdBy === 'string' ? row.createdBy : null,
        createdAt: typeof row.createdAt === 'string' ? row.createdAt : null,
        updatedAt: typeof row.updatedAt === 'string' ? row.updatedAt : null,
      };
    }),
  };
}

/**
 * One-sentence summary of what this repository's configuration does on a divergence.
 *
 * Written so the count of exceptions is never silently dropped: a repository set to
 * "overwrite" with four files held back is a materially different setup from one with none,
 * and the difference has to be visible without expanding the table.
 *
 * @param policy The repository-wide policy.
 * @param overrideCount How many files deviate from it.
 * @returns The summary sentence.
 */
export function conflictPolicySummary(policy: ConflictPolicy, overrideCount: number): string {
  const base = `${POLICY_COPY[policy].label} applies to every file in this repository`;
  if (overrideCount === 0) return `${base}.`;
  const exception =
    overrideCount === 1 ? '1 file has its own policy' : `${overrideCount} files have their own policy`;
  return `${base}, except ${exception}.`;
}

File: apiome-rest/src/app/repository_conflict_policy.py
"""
Per-repository / per-file refresh conflict policy (RAR-4.5, #3531).

The RAR-4.4 divergence guard (:mod:`app.repository_divergence_guard`) answers one
question — *has the imported version been hand-edited since?* — and applies the
safe default: hold, do not clobber. That default is right for most teams and
wrong for some. A team whose repository is the single source of truth wants the
refresh to win. A team with a review process wants the refresh parked. A team
that wants neither side to lose wants the refresh on a side branch.

This module is the policy layer that makes that choice configurable::

    diverged ──► policy: overwrite | hold-for-review | new-branch

and resolves which policy applies to a given file::

    per-file override ──► repository policy ──► hold-for-review (default)

Two decisions live here, both pure and DB-free so they can be exercised by
deterministic fixtures (the same shape as the RAR-2.2 comparator, the RAR-2.4
idempotency guard and the RAR-4.4 divergence guard):

* :func:`resolve_conflict_policy` — which policy applies, and where it came from.
* :func:`decide_conflict` — what the refresh should actually *do*, by running the
  RAR-4.4 guard under the resolved policy and mapping its verdict to a
  :class:`ConflictAction`.

Persisting the policy (``tenant_repositories.refresh_conflict_policy`` and
``apiome.repository_conflict_policy_override``, migration
``V235__repository_conflict_policy_rar_4_5.sql``) is the DAO's job; acting on the
returned :class:`ConflictAction` — superseding the version, flagging ``diverged``
and firing the RAR-5.4 notification, or creating the side branch — is the EPIC-4
dispatcher's, mirroring how RAR-4.1/4.2/4.3/4.4 deferred their wiring.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum
from typing import Optional

from .repository_divergence_guard import (
    DivergenceDecision,
    DivergencePolicy,
    DivergenceReason,
    evaluate_divergence,
)


class ConflictPolicy(str, Enum):
    """What to do when an auto-refresh meets a hand-edited (diverged) version.

    These are the stored, wire-visible policy tokens: the values match the
    ``CHECK`` constraints on ``tenant_repositories.refresh_conflict_policy`` and
    ``repository_conflict_policy_override.policy``, and are what the settings API
    accepts and returns.
    """

    #: The repository wins: the refresh supersedes the hand-edited version. The
    #: divergence is still *detected* and reported (so audit and notifications
    #: reflect reality) — it just does not stop the refresh.
    OVERWRITE = "overwrite"
    #: Hold-not-clobber (default, the RAR-4.4 behaviour): the refresh is skipped,
    #: the file is flagged ``diverged``, and a human resolves it.
    HOLD_FOR_REVIEW = "hold-for-review"
    #: Neither side loses: the current version is left untouched and the refresh
    #: lands on a new branch/version for review and merge.
    NEW_BRANCH = "new-branch"


#: The policy in force when neither the file nor the repository specifies one.
#: RAR-4.5 keeps the RAR-4.4 hold-not-clobber default: opting into a policy that
#: can lose a hand edit must be an explicit act.
DEFAULT_CONFLICT_POLICY = ConflictPolicy.HOLD_FOR_REVIEW


class ConflictPolicySource(str, Enum):
    """Where the applied policy came from (surfaced in the UI and the audit trail)."""

    #: A per-file override row (``repository_conflict_policy_override``).
    FILE = "file"
    #: The repository-wide ``tenant_repositories.refresh_conflict_policy``.
    REPOSITORY = "repository"
    #: Neither was set (or both were unrecognised) -> the built-in default.
    DEFAULT = "default"


class ConflictAction(str, Enum):
    """What the refresh executor should do with this file on this tick."""

    #: Apply the refresh to the current version, superseding it. Either nothing
    #: diverged, or the policy is :attr:`ConflictPolicy.OVERWRITE`.
    APPLY = "apply"
    #: Do not apply: skip the overwrite, flag the file ``diverged``, notify
    #: (RAR-5.4) and wait for a human.
    HOLD = "hold"
    #: Do not touch the current version: land the refresh on a new branch so both
    #: the hand edit and the upstream change survive.
    NEW_BRANCH = "new-branch"


#: How each policy maps onto the RAR-4.4 guard's policy input. ``hold-for-review``
#: and ``new-branch`` both *protect* the current version (the guard holds the
#: overwrite); they differ only in what the executor does next, which
#: :func:`decide_conflict` decides from the policy itself.
_GUARD_POLICY = {
    ConflictPolicy.OVERWRITE: DivergencePolicy.OVERWRITE,
    ConflictPolicy.HOLD_FOR_REVIEW: DivergencePolicy.HOLD,
    ConflictPolicy.NEW_BRANCH: DivergencePolicy.HOLD,
}

#: Tolerated spellings of the stored tokens. The API and the DB agree on the
#: hyphenated forms; these aliases mean a hand-written config or an older caller
#: using the RAR-4.4 guard's internal ``"hold"`` token still resolves correctly
#: instead of silently falling back to the default.
_POLICY_ALIASES = {
    "hold": ConflictPolicy.HOLD_FOR_REVIEW,
    "hold_for_review": ConflictPolicy.HOLD_FOR_REVIEW,
    "holdforreview": ConflictPolicy.HOLD_FOR_REVIEW,
    "review": ConflictPolicy.HOLD_FOR_REVIEW,
    "new_branch": ConflictPolicy.NEW_BRANCH,
    "newbranch": ConflictPolicy.NEW_BRANCH,
    "branch": ConflictPolicy.NEW_BRANCH,
    "clobber": ConflictPolicy.OVERWRITE,
}

#: Characters a git ref may not carry, collapsed to ``-`` when building a branch
#: name. Deliberately conservative: anything outside the safe set is replaced.
_UNSAFE_REF_CHARS = re.compile(r"[^A-Za-z0-9._/-]+")


def parse_conflict_policy(
    value: Optional[object],
    *,
    default: Optional[ConflictPolicy] = None,
) -> Optional[ConflictPolicy]:
    """Parse a stored/submitted policy token into a :class:`ConflictPolicy`.

    Accepts the canonical tokens (``overwrite`` / ``hold-for-review`` /
    ``new-branch``) case-insensitively, plus the underscore and legacy spellings
    in :data:`_POLICY_ALIASES`, so a value written by an older caller or typed by
    hand still resolves. Anything unrecognised — including ``None`` and blank —
    yields ``default``.

    Args:
        value: The raw policy token (typically a DB column or a request field),
            or ``None``.
        default: What to return when ``value`` is missing or unrecognised.
            Defaults to ``None`` so a caller can distinguish "not set" from "set
            to the default"; :func:`resolve_conflict_policy` relies on that.

    Returns:
        The parsed :class:`ConflictPolicy`, or ``default``.
    """
    if isinstance(value, ConflictPolicy):
        return value
    if not isinstance(value, str):
        return default
    token = value.strip().lower()
    if not token:
        return default
    try:
        return ConflictPolicy(token)
    except ValueError:
        return _POLICY_ALIASES.get(token.replace("-", "_"), default)


@dataclass(frozen=True)
class ResolvedConflictPolicy:
    """Which conflict policy applies to one file, and where it came from.

    Attributes:
        policy: The :class:`ConflictPolicy` in force for this file.
        source: The :class:`ConflictPolicySource` the policy was read from —
            ``file`` when a per-file override won, ``repository`` when the
            repository-wide setting did, ``default`` when neither was usable.
    """

    policy: ConflictPolicy
    source: ConflictPolicySource


def resolve_conflict_policy(
    *,
    file_policy: Optional[object] = None,
    repository_policy: Optional[object] = None,
    default: ConflictPolicy = DEFAULT_CONFLICT_POLICY,
) -> ResolvedConflictPolicy:
    """Resolve the conflict policy for one file (RAR-4.5).

    Applies the precedence the ticket specifies — the per-file override is the
    most specific statement of intent, the repository setting is the fallback,
    and the built-in default backs both::

        per-file override ──► repository policy ──► hold-for-review

    An unrecognised value at either level is treated as *not set* rather than as
    an error, so one bad row degrades to the next-broadest policy instead of
    failing the refresh. Because the broadest fallback is hold-not-clobber, every
    degradation path is the safe one.

    Args:
        file_policy: The per-file override token, or ``None`` when the file has no
            override row.
        repository_policy: The repository-wide token
            (``tenant_repositories.refresh_conflict_policy``), or ``None``.
        default: The policy backing both levels. Defaults to
            :data:`DEFAULT_CONFLICT_POLICY` (hold-for-review).

    Returns:
        A :class:`ResolvedConflictPolicy` carrying the policy and its source.
    """
    parsed_file = parse_conflict_policy(file_policy)
    if parsed_file is not None:
        return ResolvedConflictPolicy(
            policy=parsed_file, source=ConflictPolicySource.FILE
        )

    parsed_repo = parse_conflict_policy(repository_policy)
    if parsed_repo is not None:
        return ResolvedConflictPolicy(
            policy=parsed_repo, source=ConflictPolicySource.REPOSITORY
        )

    return ResolvedConflictPolicy(policy=default, source=ConflictPolicySource.DEFAULT)


@dataclass(frozen=True)
class ConflictOutcome:
    """What a refresh should do with one file, under the policy that applies.

    Attributes:
        action: The :class:`ConflictAction` the executor must take — apply the
            refresh, hold it, or divert it to a new branch.
        diverged: True when a manual edit was *detected*. Independent of the
            action: an ``overwrite`` policy still reports the divergence it
            clobbered, so audit and notifications reflect reality.
        reason: The RAR-4.4 :class:`DivergenceReason` behind the verdict.
        policy: The :class:`ConflictPolicy` the decision was made under.
        source: Where that policy came from (:class:`ConflictPolicySource`).
        branch_name: The branch the refresh should land on, set only when
            ``action`` is :attr:`ConflictAction.NEW_BRANCH` and enough context was
            supplied to name one; ``None`` otherwise.
    """

    action: ConflictAction
    diverged: bool
    reason: DivergenceReason
    policy: ConflictPolicy
    source: ConflictPolicySource
    branch_name: Optional[str] = None


def refresh_branch_name(
    *,
    base_branch: Optional[str],
    path: Optional[str] = None,
    commit_sha: Optional[str] = None,
) -> Optional[str]:
    """Build the branch name a ``new-branch`` refresh lands on.

    Deterministic — the same inputs always name the same branch — so a refresh
    that runs twice for one commit targets one branch rather than accumulating
    near-duplicates. The shape is::

        apiome-refresh/<base branch>/<file stem>-<short sha>

    Unsafe ref characters are collapsed to ``-`` and the sha is shortened to 12
    characters, which is unambiguous in practice and keeps the name readable.

    Args:
        base_branch: The branch the file was imported from. ``None``/blank means
            the caller has no branch context and no name can be built.
        path: The repository path of the refreshed file; its file stem
            disambiguates two files refreshed from the same commit. Optional.
        commit_sha: The source commit driving the refresh (RAR-2.1). Optional;
            omitted when the caller has no commit signal.

    Returns:
        The branch name, or ``None`` when ``base_branch`` is missing/blank (the
        caller then has to name the branch itself).
    """
    base = (base_branch or "").strip().strip("/")
    if not base:
        return None

    segments = [_safe_ref_segment(base)]

    leaf = ""
    if path:
        tail = str(path).strip().rstrip("/").rsplit("/", 1)[-1]
        leaf = _safe_ref_segment(tail.rsplit(".", 1)[0])
    sha = _safe_ref_segment((commit_sha or "").strip())[:12]

    suffix = "-".join(part for part in (leaf, sha) if part)
    if suffix:
        segments.append(suffix)

    return "apiome-refresh/" + "/".join(part for part in segments if part)


def _safe_ref_segment(value: str) -> str:
    """Reduce a string to characters that are safe inside a git ref segment."""
    cleaned = _UNSAFE_REF_CHARS.sub("-", value).strip("-./")
    return cleaned


def decide_conflict(
    *,
    post_import_checksum: Optional[str],
    current_checksum: Optional[str],
    file_policy: Optional[object] = None,
    repository_policy: Optional[object] = None,
    base_branch: Optional[str] = None,
    path: Optional[str] = None,
    commit_sha: Optional[str] = None,
) -> ConflictOutcome:
    """Decide what a refresh does with one file, under its conflict policy (RAR-4.5).

    Resolves the applicable policy (:func:`resolve_conflict_policy`), runs the
    RAR-4.4 divergence guard under it (:func:`~app.repository_divergence_guard.
    evaluate_divergence`), and maps the verdict to the action the executor takes:

    ==================  ================  ===========================================
    Divergence          Policy            Action
    ==================  ================  ===========================================
    none                any               ``apply``
    detected            ``overwrite``     ``apply``   (divergence reported, not held)
    detected            ``hold-for-review``  ``hold``  (default: no clobber)
    detected            ``new-branch``    ``new-branch`` (current version untouched)
    ==================  ================  ===========================================

    Divergence detection itself is unchanged from RAR-4.4: no baseline fails open,
    an identical checksum applies, and a missing current checksum with a baseline
    present counts as a difference.

    Args:
        post_import_checksum: Content checksum captured right after the original
            import (the snapshot baseline), or ``None`` when none was captured.
        current_checksum: Content checksum of the version as it stands now.
        file_policy: The per-file override token, or ``None``.
        repository_policy: The repository-wide policy token, or ``None``.
        base_branch: The branch the file was imported from, used to name the
            branch when the action is ``new-branch``. Optional.
        path: The repository path of the file, used in the branch name. Optional.
        commit_sha: The source commit driving the refresh, used in the branch
            name. Optional.

    Returns:
        A :class:`ConflictOutcome` describing the action, the divergence verdict,
        the policy applied and its source.
    """
    resolved = resolve_conflict_policy(
        file_policy=file_policy, repository_policy=repository_policy
    )
    decision: DivergenceDecision = evaluate_divergence(
        post_import_checksum=post_import_checksum,
        current_checksum=current_checksum,
        policy=_GUARD_POLICY[resolved.policy],
    )

    if not decision.should_hold:
        # Either nothing diverged, or the policy is OVERWRITE and the refresh is
        # allowed to supersede the hand edit. Both apply; `diverged` still carries
        # whether an edit was clobbered.
        action = ConflictAction.APPLY
        branch_name = None
    elif resolved.policy is ConflictPolicy.NEW_BRANCH:
        # The current version is protected (the guard held the overwrite) and the
        # refresh is diverted rather than dropped.
        action = ConflictAction.NEW_BRANCH
        branch_name = refresh_branch_name(
            base_branch=base_branch, path=path, commit_sha=commit_sha
        )
    else:
        action = ConflictAction.HOLD
        branch_name = None

    return ConflictOutcome(
        action=action,
        diverged=decision.diverged,
        reason=decision.reason,
        policy=resolved.policy,
        source=resolved.source,
        branch_name=branch_name,
    )
